### ISSUE TITLE
a[i] in steps for failure printouts ...

### ISSUE THREAD

**Issue by jleonard-r7** [quickstrom/quickstrom#70]
This is more of a question than an issue (but perhaps may be an issue).

Would "click a[i]" be equivalent to the following JavaScript?

```
document.querySelectorAll("a")[i].click();
```

It doesn't seem to be lining up that way in my testing. And if it isn't, then perhaps some better output for steps is called for so that manual reproduction / diagnosis of failures would be possible...

**Comment by owickstrom** [comment]
Ah, yes this is perhaps misleading. It would be the same as your line of JS **if all `a` elements returned by the query are deemed clickable**. So, in the following case:

```html
<button>A</button>
<button disabled>B</button>
<button>C</button>
```
Then `Click button[1]` would refer to button `C`.

**Comment by owickstrom** [comment]
Maybe better to be verbose and print something like `5th clickable element`.

**Comment by owickstrom** [comment]
Related, I'm going to add hover support to actions in the HTML report also. Then you can get a marker on the screenshot which element was interacted with by an action.

**Comment by jleonard-r7** [comment]
What is the predicate (in JS pseudo code preferable) that determines if something is “clickable”? Is it just !disabled or more complex?

Also, is the HTML reporting in a usable state as of yet? If so, are there docs on how to use it yet?

**Comment by owickstrom** [comment]
Unfortunately it's more complex, in order to prevent actions that WebDriver will later reject. So it checks if elements are visible, enabled, if inputs are focused, etc. Quickstrom currently lacks but needs to add a check if elements are inside the viewport, too.

Yeah, it's not quite "done" but should be usable. You'll find the docs at https://docs.quickstrom.io/topics/reporters.html#html.

**Comment by jleonard-r7** [comment]
Ah, so the following error(s) are likely emanating from the elements being out of viewport?

```
2021-01-22 00:43:10 ERROR  Error Response: 400 Bad Request
{
    "data": null,
    "error": 400,
    "stacktrace": "#0 0x556098a7eaa9 <unknown>\n",
    "message": "element not interactable\n  (Session info: headless chrome=85.0.4183.83)"
}
```

**Comment by owickstrom** [comment]
Yes, that's what I suspect. Quickstrom should exclude those actions with some predicate.

**Comment by jleonard-r7** [comment]
I have been unable to get Quickstrom to generate HTML reports. I added `--reporter=html` and `--html-report-directory` to its arguments but no files get produced in the provided directory. If I try with the console reporter, I get the usual output.

**Comment by owickstrom** [comment]
Might it be that the directory already exists? It then refuses to overwrite it, so you have to provide a new directory (like `mktemp -u -d`) or delete the old one.

**Comment by jleonard-r7** [comment]
Unfortunately that wasn't it. I tried with and without the directory pre-existing.

FWIW, here are the last few log lines from my Quickstrom run (if that helps any):

```
...
2021-02-09 20:21:09 ERROR  Error Response: 400 Bad Request
{
    "data": null,
    "error": 400,
    "stacktrace": "#0 0x556936045aa9 <unknown>\n",
    "message": "element not interactable\n  (Session info: headless chrome=85.0.4183.83)"
}
Running shrunk test at level 1.
Running shrunk test at level 1.
Running shrunk test at level 2.
Running shrunk test at level 2.
Running shrunk test at level 2.
Running shrunk test at level 3.
Running shrunk test at level 3.
Running shrunk test at level 3.
Running shrunk test at level 3.
Running shrunk test at level 4.
Running shrunk test at level 4.
Running shrunk test at level 4.
Running shrunk test at level 4.
Running shrunk test at level 4.
Running shrunk test at level 4.
2021-02-09 20:38:33 ERROR  Error Response: 400 Bad Request
{
    "data": null,
    "error": 400,
    "stacktrace": "#0 0x56207e25aaa9 <unknown>\n",
    "message": "element not interactable\n  (Session info: headless chrome=85.0.4183.83)"
}
Running shrunk test at level 4.
Running shrunk test at level 5.
Running shrunk test at level 5.
Running shrunk test at level 5.
Running shrunk test at level 5.

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

ubuntu@ip-redacted:~/src/my-app$ 
```

**Comment by owickstrom** [comment]
OK. What command did you run? Would it be possible to post the entire log?

**Comment by jleonard-r7** [comment]
This is the command:

```
docker run --rm --network quickstrom --volume="$PWD:/web-cljs" quickstrom/quickstrom:latest quickstrom check --reporter=html --html-report-directory=./quickstrom/report --max-trailing-state-changes=10 --trailing-state-change-timeout=250 --webdriver-host=webdriver --webdriver-path=/wd/hub --browser=chrome --tests=5 /web-cljs/quickstrom/specs/app.spec.purs https://oursite.com/login 2>&1 | tee output.txt
```

I can post the log a bit later on.

**Comment by owickstrom** [comment]
Ah, I think I know what is. You're running `quickstrom` through Docker, and the HTML report is written inside the docker container. `./quickstrom` isn't a mounted volume, right? So it writes the HTML report to a directory only inside the container. If you add a volume mount and use the absolute path for `--html-report-directory`, does that work?

**Comment by jleonard-r7** [comment]
Ah yep, that was it. Using an absolute path into the already mounted $PWD works.

**Comment by owickstrom** [comment]
Nice!

**Comment by owickstrom** [comment]
Closing this issue now. Feel free to open a new one if the printouts are problematic.

**Comment by jleonard-r7** [comment]
This issue pertains to the CLI reports, not the HTML ones. Are you suggesting that CLI reports are deprecated ?

**Comment by jleonard-r7** [comment]
And actually this issue still applies to the HTML reports (which still list clicks as, e.g., "a[i] clicked" without any way to know which a was actually clicked). The only difference is that there is indeed more circumstantial evidence (i.e., the screenshots) but it isn't always possible to infer what happened (or even when it is, it can be difficult).

**Comment by owickstrom** [comment]
Sorry, got a bit derailed by our discussion. OK, so let's add some specific new issues, if these make sense:

* Show which if the *N* possible candidates were taken (e.g. "5th out of 7 candidate elements") for both CLI and HTML reports
* Add a highlight-on-hover for the actions in the HTML report

**Comment by jleonard-r7** [comment]
That sounds awesome. Two things though:

For the CLI solution, I'm not sure how much that would help. Perhaps we can think about a "deeper" fix or some way to integrate w/ screenshots/highlights for the CLI as well.

For HTML, that would be awesome. I think we will need to be mindful of the positions of the clicked elements such that we can "scroll" them into view (via an animation) on hover. (Either that or resize the viewport to take up the full size of the render). The former is probably more impressive from a pure wow factor point of view but the latter may end up being more practical / useful (as the distinguishing/relevant bits may be off screen [or far away from the click]). Also, the "resize viewport" option could be done globally independently of clicks [for the same reason(s) of usefulness/practicality].

**Comment by owickstrom** [comment]
> For the CLI solution, I'm not sure how much that would help. Perhaps we can think about a "deeper" fix or some way to integrate w/ screenshots/highlights for the CLI as well.

Yeah I've thought about writing screenshot files in that case to, but with "static" highlights already added (using something like ffmpeg or GraphicsMagick to draw rectangles). That would work for if only highlighting the action, but perhaps not highlight all queried elements.

Also, if we write a directory of screenshot files, then we're sort of half way to the HTML anyway. I guess it depends on how you want to consume the information. I'm starting to lean towards the console reporter not being a very useful troubleshooting tool for failed checks.

Another option is writing a JSON file (like the `.jsonp` file in the HTML report directory) with all information in some standardized format, and then writing tooling for displaying that report data. Possibly as separate projects. This is what I'd like to do for https://app.quickstrom.io.

> For HTML, that would be awesome. I think we will need to be mindful of the positions of the clicked elements such that we can "scroll" them into view (via an animation) on hover. (Either that or resize the viewport to take up the full size of the render). The former is probably more impressive from a pure wow factor point of view but the latter may end up being more practical / useful (as the distinguishing/relevant bits may be off screen [or far away from the click]). Also, the "resize viewport" option could be done globally independently of clicks [for the same reason(s) of usefulness/practicality].

Right. Currently the screenshots are scaled down to fit horizontally, but not vertically. I had planned to instead display them at a 1:1 size with overflow, and perhaps automatically scroll. If I understand you correctly, you mean we could supply two modes which you could toggle between? Like "100%" or "Fit to screen".

And automatic scrolling could be cool, but also potentially very annoying. Maybe that should also be a user setting.

**Comment by jleonard-r7** [comment]
Yes for CLI, the JSON file + separate renderer sounds good although I agree it is starting to approach the HTML approach at that point. 

Regarding HTML, actually rather than two separate modes, it just occurred to me that we could support both simultaneously if we merely allow the screenshots to themselves be scrollable. It would effectively give the user the impression of having miniature browser windows inside the page. And they could be “pre-scrolled” to the relevant element for the action in any transition page. Then if the user were to scroll away, a hover over the action could once again “re-scroll” it back into view.
